presto-main/src/main/java/com/facebook/presto/execution/buffer/SpoolingOutputBuffer.java

```diff
@@ -0,0 +1,534 @@
+            this.threshold = threshold;
+
+            context = new TempDataOperationContext(Optional.empty(), taskId.getQueryId().toString(), Optional.empty(), new Identity("spooling-buffer", Optional.empty()));
+        }
+        catch (Exception e) {
+            throw new IllegalArgumentException();
+        }
+    }
+
+    @Override
+    public OutputBufferInfo getInfo()
+    {
+        return new OutputBufferInfo(
+                "SPOOLING",
+                state.get(),
+                state.get().canAddBuffers(),
+                state.get().canAddPages(),
+                totalBufferedBytes.get(),
+                totalBufferedPages.get(),
+                totalRowsAdded.get(),
+                totalPagesAdded.get(),
+                ImmutableList.of()
+        );
+    }
+
+    @Override
+    public boolean isFinished()
+    {
+        return state.get() == BufferState.FINISHED;
+    }
+
+    @Override
+    public double getUtilization()
+    {
+        return 0;
+    }
+
+    @Override
+    public boolean isOverutilized()
+    {
+        return false;
+    }
+
+    @Override
+    public void addStateChangeListener(StateMachine.StateChangeListener<BufferState> stateChangeListener)
+    {
+        state.addStateChangeListener(stateChangeListener);
+    }
+
+    @Override
+    public void setOutputBuffers(OutputBuffers newOutputBuffers)
+    {
+        requireNonNull(newOutputBuffers, "newOutputBuffers is null");
+        checkArgument(outputBuffers.getType() == OutputBuffers.BufferType.SPOOLING, "Invalid output buffers type");
+        checkArgument(outputBuffers.isNoMoreBufferIds(), "invalid noMoreBufferIds");
+
+        if (state.get().isTerminal() || outputBuffers.getVersion() >= newOutputBuffers.getVersion()) {
+            return;
+        }
+
+        outputBuffers.checkValidTransition(newOutputBuffers);
+    }
+
+    @Override
+    public ListenableFuture<BufferResult> get(OutputBuffers.OutputBufferId bufferId, long token, DataSize maxSize)
+    {
+        requireNonNull(bufferId, "outputBufferId is null");
+        checkArgument(bufferId.getId() == 0, "Invalid buffer id");
+        checkArgument(maxSize.toBytes() > 0, "maxSize must be at least 1 byte");
+        acknowledge(bufferId, token);
+        long currentSequenceId = sequenceId.get();
+
+        // token is for a page before the current page position
+        if (token < currentSequenceId) {
+            return immediateFuture(emptyResults(taskInstanceId, token, false));
+        }
+
+        // tells client buffer is complete
+        if (noMorePages.get() && handleRangeMap.asMapOfRanges().isEmpty() && pages.isEmpty()) {
+            return immediateFuture(emptyResults(taskInstanceId, token, true));
+        }
+
+        checkArgument(currentSequenceId == token, "Invalid token");
+
+        try {
+            long maxBytes = maxSize.toBytes();
+            long bytes = 0;
+            List<SerializedPage> result = new LinkedList<>();
```

Comment:
I would use an immutable list builder here.  Search the codebase for example of `ImmutableList.Builder` for examples.